From 095ccd28fd987c9dbed131a0d8c6c62ca222e855 Mon Sep 17 00:00:00 2001
From: Zhang Zekun <zhangzekun11@huawei.com>
Date: Mon, 20 Mar 2023 13:35:40 +0000
Subject: [PATCH] Revert: "mm: sharepool: Charge Buddy hugepage to memcg"

Charge buddy hugepage to memcg could cause some unexpected
errors, revert it to fix this.

Signed-off-by: Zhang Zekun <zhangzekun11@huawei.com>
---
 mm/hugetlb.c             |  2 --
 mm/share_pool.c          | 13 ++-----------
 mm/share_pool_internal.h | 26 --------------------------
 3 files changed, 2 insertions(+), 39 deletions(-)
 delete mode 100644 mm/share_pool_internal.h

diff --git a/mm/hugetlb.c b/mm/hugetlb.c
index 1e46507eb..aa572bd6f 100644
--- a/mm/hugetlb.c
+++ b/mm/hugetlb.c
@@ -46,7 +46,6 @@
 #include <linux/dynamic_hugetlb.h>
 #include "internal.h"
 #include "hugetlb_vmemmap.h"
-#include "share_pool_internal.h"
 
 int hugetlb_max_hstate __read_mostly;
 unsigned int default_hstate_idx;
@@ -1628,7 +1627,6 @@ void free_huge_page(struct page *page)
 		h->resv_huge_pages++;
 
 	if (HPageTemporary(page)) {
-		sp_memcg_uncharge_hpage(page);
 		remove_hugetlb_page(h, page, false);
 		spin_unlock_irqrestore(&hugetlb_lock, flags);
 		update_and_free_page(h, page, true);
diff --git a/mm/share_pool.c b/mm/share_pool.c
index 73282ec58..15cffa829 100644
--- a/mm/share_pool.c
+++ b/mm/share_pool.c
@@ -4050,7 +4050,6 @@ vm_fault_t sharepool_no_page(struct mm_struct *mm,
 	bool new_page = false;
 	int err;
 	struct sp_area *spa;
-	bool charge_hpage;
 
 	spa = vma->vm_private_data;
 	if (!spa) {
@@ -4065,15 +4064,12 @@ retry:
 		if (idx >= size)
 			goto out;
 
-		charge_hpage = false;
 		page = alloc_huge_page(vma, haddr, 0);
 		if (IS_ERR(page)) {
 			page = hugetlb_alloc_hugepage_vma(vma, haddr,
 					HUGETLB_ALLOC_BUDDY | HUGETLB_ALLOC_NORECLAIM);
 			if (!page)
 				page = ERR_PTR(-ENOMEM);
-			else if (!PageMemcgKmem(page))
-				charge_hpage = true;
 		}
 		if (IS_ERR(page)) {
 			ptl = huge_pte_lock(h, mm, ptep);
@@ -4087,12 +4083,6 @@ retry:
 			goto out;
 		}
 
-		if (charge_hpage && mem_cgroup_charge(page, vma->vm_mm, GFP_KERNEL)) {
-			put_page(page);
-			ret = vmf_error(-ENOMEM);
-			goto out;
-		}
-
 		__SetPageUptodate(page);
 		new_page = true;
 
@@ -4125,8 +4115,9 @@ retry:
 
 	spin_unlock(ptl);
 
-	if (new_page)
+	if (new_page) {
 		SetPagePrivate(&page[1]);
+	}
 
 	unlock_page(page);
 out:
diff --git a/mm/share_pool_internal.h b/mm/share_pool_internal.h
deleted file mode 100644
index 74ebf4739..000000000
--- a/mm/share_pool_internal.h
+++ /dev/null
@@ -1,26 +0,0 @@
-/* SPDX-License-Identifier: GPL-2.0 */
-/*
- * Copyright (C) Huawei Technologies Co., Ltd. 2023. All rights reserved.
- */
-
-#ifndef SHARE_POOL_INTERNAL_H
-#define SHARE_POOL_INTERNAL_H
-
-#include <linux/hugetlb.h>
-#include <linux/memcontrol.h>
-
-#ifdef CONFIG_ASCEND_SHARE_POOL
-static inline void sp_memcg_uncharge_hpage(struct page *page)
-{
-	if (!sp_is_enabled())
-		return;
-
-	mem_cgroup_uncharge(page);
-}
-#else
-static inline void sp_memcg_uncharge_hpage(struct page *page)
-{
-}
-#endif
-
-#endif
-- 
2.17.1

